### PR TITLE
`resolve_tc c` solves typeclasses appearing in `c`

### PR DESCRIPTION
**Pull request by maximedenes**
Hopefully this can help Iris to not depend on #6583 (see discussion there).

@robbertkrebbers and @RalfJung, would you mind letting me know if this fixes your idiom which depended on `apply` resolving unrelated goals? Thanks!

- [x] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

**Comment by RalfJung** [comment]
I entirely lost context, so I am not sure what should be tested here. There are a couple of testcases in https://github.com/coq/coq/issues/6583 but I assume you tried those?

I guess we could change the `refine` [mentioned here](https://gitlab.mpi-sws.org/iris/iris/-/merge_requests/330#note_41327) back into an `apply` and see what happens? Not sure if that makes any sense... @robbertkrebbers might know better.

**Comment by RalfJung** [comment]
Even with this branch, if I change the `refine` back to an `apply` I get build failures in Iris:
```
File "/home/r/Dokumente/Unisachen/iris/coq.2/tests/proofmode_iris.v", line 238, characters 17-35:
Error:
Unable to satisfy the following constraints:
In environment:
Σ : gFunctors
invG0 : invG Σ
cinvG0 : cinvG Σ
na_invG0 : na_invG Σ
A : cmraT
inG0 : inG Σ A
γ : gname
x, x' : A
H : x' ≼ x

?Hin : "inG ?Σ ?A"
```
However, I do not know if this is the right way to test that change.

**Comment by robbertkrebbers** [comment]
This is very interesting. Thanks for working on this @maximedenes.

So, just to get clear what you've done: This MR defines a new tactic `resolve_tc` that can be used to implement step (3) in https://github.com/coq/coq/issues/6583#issuecomment-582632499 ?

**Comment by robbertkrebbers** [comment]
@RalfJung In https://github.com/coq/coq/issues/6583#issuecomment-582632499 I pointed out that Iris actually relies on a weird behavior of the legacy `apply` tactic. This made it impossible to switch to `refine` variants for some parts of the proof mode code. (And thus, that would make it impossible to port the proof mode to a hypothetical new version of Coq that uses the new unification algorithm by default).

The MR seems to provide a way that makes it possible to implement the proof mode solely using `refine`, so without relying on the legacy `apply` tactic.

**Comment by RalfJung** [comment]
Oh I see, I misunderstood. This does not change the behavior of existing tactics but adds a new tactic that would be useful for us.

I'll have to leave testing to Robbert then. :)

**Comment by maximedenes** [comment]
> So, just to get clear what you've done: This MR defines a new tactic `resolve_tc` that can be used to implement step (3) in [#6583 (comment)](https://github.com/coq/coq/issues/6583#issuecomment-582632499) ?

Yes, at least that's what I tried to do ;)

> Oh I see, I misunderstood. This does not change the behavior of existing tactics but adds a new tactic that would be useful for us.

Sorry, I should have put more context. The problem is that Iris is *depending* on the bug that you'd like us to fix ;) Hopefully, the tactic defined in this PR should be enough, so that we can then fix the bug without breaking Iris.

**Comment by maximedenes** [comment]
@robbertkrebbers if your testing shows that this tactic suits your needs, I'll document it and polish the patch. Feel free to suggest a better name for the tactic as well, or any other improvements.

**Comment by robbertkrebbers** [comment]
@maximedenes I have been experimenting with this, and while the tactic seems helpful, I cannot quite use it for the code in the current state. I have code like:

```
apply lemma;
  [create term t with evars
  |do other stuff];
(* now, after "do other stuff" is done, I want to solve all remaining evars in t *)
```

The problem is that the scope of `t` is restricted, and I cannot refer to it after the `apply`.

I'm trying to rewrite the code to avoid this, but it's complicated.

If you have any immediate suggestions, they are most welcome. Otherwise I will keep on trying, and post once I know more.

**Comment by ppedrot** [comment]
It's fairly easy to add a continuation to this tactic though, so that it allows to implement your scope-building pattern.

**Comment by maximedenes** [comment]
> It's fairly easy to add a continuation to this tactic though, so that it allows to implement your scope-building pattern.

@ppedrot if you have a minute, would you mind adding it to the PR? It is an experiment anyway. Thanks!

**Comment by robbertkrebbers** [comment]
PS: Another fun problem in the code in https://github.com/coq/coq/pull/13071#issuecomment-704836525:

It is often the case that "do other stuff" solves all goals, and then `resolve_tc` is not called. I managed to hack around that by adding a dummy premise `True` to `lemma`.

**Comment by robbertkrebbers** [comment]
> > It's fairly easy to add a continuation to this tactic though, so that it allows to implement your scope-building pattern.
> 
> @ppedrot if you have a minute, would you mind adding it to the PR? It is an experiment anyway. Thanks!

That would be great indeed, would be curious to see how that looks like and if it helps.

**Comment by maximedenes** [comment]
> The problem is that the scope of `t` is restricted, and I cannot refer to it after the `apply`.

@robbertkrebbers do you mind elaborating? In #6583, you wrote that the idiom used in Iris was:

```coq
Tactic Notation "foo" uconstr(c) :=
  let c := open_constr(c) in                                (* step 1 *)               
  perform tactics;                                          (* step 2 *)
  call TC search on all remaining evars created in step 1   (* step 3 *)
```

Which seems easy to transform into:

```coq
Tactic Notation "foo" uconstr(c) :=
  let c := open_constr(c) in                                (* step 1 *)               
  perform tactics;                                          (* step 2 *)
  resolve_tc c
```

Having `resolve_tc` take a continuation seems indeed useful when "do other stuff" solves all goals, but I don't get exactly what the scoping problem is.

**Comment by robbertkrebbers** [comment]
The actual code turned out to be a bit more complicated, unfortunately...

Maybe I should try to explain what we try to achieve, instead of how it's currently implemented. I actually think it's not really Iris specific, so we might be able to come up with a general-purpose solution.

Basically, many of the Iris tactics (for example, `iApply`) take lemmas in the Iris logic, which are of the form:

```coq
lem : ∀ x1 .. xn, (P ⊢ Q)
```
Here, `⊢ : iProp → iProp → Prop` is the entailment relation of Iris. An important thing to point out about these lemmas is that the universal quantification `∀ x1 .. xn` happens at the Coq level (the `∀` lives in Coq's `Prop`). A concrete examples of such a lemma is:

```coq
own_update :
  ∀ (Σ : gFunctors) (A : cmraT) (i : inG Σ A) (γ : gname) (a a' : A), (a ~~> a') → (own γ a ⊢ |==> own γ a')
Arguments own_update {Σ A} [i]
```
When a user gives a lemma to an Iris tactic like `iApply`, similarly to using Coq's `apply` tactic, it typically does not instantiate all of the arguments of the lemma. The user typically only specifies a prefix of the arguments, and for some of the arguments it does specify, it would use a `_` so that these could be inferred. To support the latter, tactics like `iApply` take an `open_constr` as their argument. To deal with the fact that the user only supplies a prefix of the arguments, these tactics perform a loop, in which we apply the lemma until an Iris entailment is reached. In this loop, we create explicit goals for arguments that are Props (for example, `a ~~> a'` in the above example), and evars for anything else.

Then, once we have obtained the fully applied lemma, which then has type `P ⊢Q` (where `P` and `Q` contain a bunch of evars), we try to unify the `Q` against the Iris proof goal. As part of this process, many of the evars that we created when applying the lemma get instantiated. But at the end of all this, it's often the case that some evars are still left uninstantiated. These we wish to solve by calling TC search.

The implementation of the current tactic that performs the loop to instantiate arguments is implemented here: https://gitlab.mpi-sws.org/iris/iris/-/blob/master/theories/proofmode/ltac_tactics.v#L808 The core part is `iIntoEmpValid`, but this one is performed in a subtactic which is not scoped to the entire tactic. Hence the problem I described above. Probably, what I am doing here more of a consequence of getting something to work despite the crazy Ltac semantics, rather than an ideal implementation.

I tried implementing the loop as something like:

```coq
Ltac apply_all_arguments lem tac :=
  first
  [(* Prop argument, create explicit goal using assert *)
   let P := fresh in
   evar (P : Prop);
   let P' := eval unfold P in P in
   clear P;
   let HP := fresh in
   assert P' as HP;
     [|let lem := open_constr:(lem HP) in
       clear HP;
       apply_all_arguments lem tac]
  |(* Other argument, e.g., a variable, create evar *)
   let lem := open_constr:(lem _) in
   apply_all_arguments lem tac
  |(* We have reached the base case *)
   tac lem;
   resolve_tc lem].
```

Here, `apply_all_arguments` takes a term `lem` (with holes), keeps on applying it, and calls `tac` on the result. This doesn't quite work because a.) whenever `tac` fails, it will backtrack on the other two cases of the `first` b.) when `tac` solves all goals, `resolve_tc` will never be called.

So, basically what might help is to have a version of `apply_all_arguments` that does not suffer from these issues. I don't know how to implement this in Ltac.

**Comment by robbertkrebbers** [comment]
This problem recently stroke us again in Iris. (See https://mattermost.mpi-sws.org/iris/pl/1ufy9cmimi8hfeiczyhx4txsmy). It would be good to have a phone call with the parties involved to discuss this MR and how we can move forward?

@RalfJung @maximedenes @SkySkimmer : Would Wed 25 Nov after 15:00 work?

**Comment by RalfJung** [comment]
See https://gitlab.mpi-sws.org/iris/iris/-/merge_requests/582 for the fix required for now -- without this, several Iris reverse dependencies fail due to TC resolution happening too early.

> Would Wed 25 Nov after 15:00 work?

:+1:

**Comment by SkySkimmer** [comment]
>@RalfJung @maximedenes @SkySkimmer : Would Wed 25 Nov after 15:00 work?

OK for me

**Comment by SkySkimmer** [comment]
ping @maximedenes

**Comment by maximedenes** [comment]
Sorry, I missed the ping. If the slot is still ok for you, I'll be available.

**Comment by SkySkimmer** [comment]
Turns out we can remove the enter but it's insufficient as we need to focus to parse the argument as a constr.
So we still need the continuation style version.

**Comment by JasonGross** [comment]
What if you have `resolve_tc` take a var instead?  I'm not sure exactly how to pull a constr out of an Ltac variable, but conceptually this fixes the focusing issue.

**Comment by SkySkimmer** [comment]
> I'm not sure exactly how to pull a constr out of an Ltac variable

I don't know either
Maybe @ppedrot knows

**Comment by ppedrot** [comment]
You need to fetch it from the `ist` argument of the tactic which is implicitly available inside a TACTIC EXTEND, and cast it as a constr using Taccoerce.

**Comment by SkySkimmer** [comment]
I tried again (see last commit) but it still focuses somewhere.

**Comment by JasonGross** [comment]
I'm confused.  Does preident focus?  Is there a way to figure out which part results in focussing?

**Comment by SkySkimmer** [comment]
@ppedrot so how do we make a non-focusing tactic?

**Comment by ppedrot** [comment]
Your primitive tactic should already be non-focussing. I will have a deeper look again when I have a bit of time.

**Comment by maximedenes** [comment]
@SkySkimmer @ppedrot do you think you could have a new look, to identify where focusing occurs?

**Comment by SkySkimmer** [comment]
This preident stuff did not work even a little so I reverted to the previous commit.

**Comment by SkySkimmer** [comment]
Here is a version which doesn't focus, by adding resolve_tc to the core tactics (tactic_expr).

**Comment by maximedenes** [comment]
IIRC we were missing feedback on the last implementation attempt. If you are willing to test it a bit, we can certainly reopen the PR. @RalfJung @robbertkrebbers is this still of interest to you?

**Comment by robbertkrebbers** [comment]
Yes, https://github.com/coq/coq/issues/6583 is still something we run in regularly, and this pull request might provide a way to make progress on that.

However, I missed that this pull request was updated. I will try to experiment with it.

**Comment by robbertkrebbers** [comment]
Could someone help me with rebasing this pull request on master? I tried it myself, and after fixing a conflict in `pptactic.ml`, I end up with a build failure (unrelated to the conflict):

```
File "tactics/class_tactics.ml", line 1330, characters 14-35:
1330 |   let debug = get_typeclasses_debug () in
                     ^^^^^^^^^^^^^^^^^^^^^
Error: Unbound value get_typeclasses_debug
Hint: Did you mean set_typeclasses_debug?
File "plugins/ltac/tactic_debug.ml", lines 350-381, characters 11-28:
350 | ...........match tac with
351 |   | TacAtom _ -> "TacAtom"
352 |   | TacThen _ -> "TacThen"
353 |   | TacDispatch _ -> "TacDispatch"
354 |   | TacExtendTac _ -> "TacExtendTac"
...
378 |   | TacArg _ -> "TacArg"
379 |   | TacSelect _ -> "TacSelect"
380 |   | TacML _ -> "TacML"
381 |   | TacAlias _ -> "TacAlias"
Error (warning 8 [partial-match]): this pattern-matching is not exhaustive.
Here is an example of a case that is not matched:
TacResolveTC _
make: *** [Makefile:145: world] Error 1
```

**Comment by SkySkimmer** [comment]
Rebased at 5039945f6f8ebe2e37a0c13e407364de95fecf5d https://github.com/SkySkimmer/coq/tree/resolve-tc-in-term

**Comment by robbertkrebbers** [comment]
Thanks for rebasing. This pull request works like a charm!

By inserting two `resolve_tc`s at the right places, I managed to remove all uses of the "legacy" `apply` in the Iris Proof Mode and replace them all by `notypeclasses refine`. See https://gitlab.mpi-sws.org/iris/iris/-/commits/robbert/resolve_tc I tested it on Iris itself, iris examples, Actris, and LambdaRust. All of them go through as expected.

So, I think this PR should be reopened and merged :).

File: test-suite/success/resolve_tc.v [hunk 1]
```diff
@@ -0,0 +1,13 @@
+Class C := c {}.
+Local Existing Instance c.
+
+Lemma foo : C.
+Proof.
+  let x := open_constr:(_:C) in resolve_tc x; exact x.
+Qed.
+
+(* TODO figure out how to avoid focusing *)
```
Comment (robbertkrebbers) [inline test-suite/success/resolve_tc.v:9]: I think this TODO can be removed.

**Comment by SkySkimmer** [comment]
Changed to define `resolve_tc` in Ltac2 instead of Ltac1. Ltac2 gives more control of focusing to the user so we don't need to add a parsing rule or special AST node.

cf test for how to call it from Ltac1.

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
@coqbot merge now

**Comment by robbertkrebbers** [comment]
Thanks all for getting this merged!

**Comment by jfehrle** [comment]
@maximedenes Would you add documentation for the new tactic?  Without it, how do you expect users to discover it?

**Comment by ppedrot** [comment]
Ltac2 tactics are directly documented in the v file as a coqdoc comment.

File: test-suite/success/resolve_tc.v [hunk 2]
```diff
@@ -0,0 +1,26 @@
+Class C := c {}.
+Local Existing Instance c.
+
+Require Ltac2.Ltac2.
+
+Ltac resolve_tc := ltac2:(x |- Std.resolve_tc (Option.get (Ltac1.to_constr x))).
```
Comment (jfehrle) [inline test-suite/success/resolve_tc.v:6]: Shouldn't this line should be included in the Std.v file so users don't have to reinvent it all the time?

**Comment by jfehrle** [comment]
Having the description _only_ in coqdoc comments is completely inadequate.

The tactic should be mentioned in the typeclasses section of the manual, the obvious place for users to look for typeclass-related tactics.  It should be identified as requiring the Ltac2 plugin and `resolve_tc` would then appear in the documentation index just like other tactics.  It can be a terse description for now. 
 Being in the standard documentation also makes it possible to hyperlink to the tactic definition and makes it possible to include executed examples through the `coqtop` directive.

The Coqdoc indexes doesn't list symbols defined with `Ltac2 @ external foobar`.  Furthermore, there's nothing in the .v files that obviously distinguishes tactics such as `resolve_tc` from non-tactic routines such as `Message.print`.

FWIW, the vast majority of items defined in the .v files have any doc comments.  That's not so helpful

Coqdoc could be enhanced to make it much better, but in its current form it's not very useful.
